### DISCUSSION TITLE
Can Supabase handle complex logic without Backend?

### DISCUSSION THREAD

**Discussion by nguyenthanhtien** [supabase/supabase#22345] [Questions]
Our team want using Supabase instead of BE. Supabase-js with NextJS, can using that to create CRM serve for 100k CCU. Supabase can support for that? My tech stack of our team consider while develop:

BE: supabase deploy in Azure Kubernetes Service.
FE: Nextjs/React

**Comment by truexor** [comment]
Hi, according to [Supabase docs](https://supabase.com/docs/reference/javascript/select), the max you can request from the Supabase through any endpoint is **1000 entries**. 
Supabase uses asynchronous tools to listen and serve data on request so the only factor to consider is the 1000 rate limit.

![image](https://github.com/supabase/supabase/assets/96081625/59f218b8-8182-4083-8ab4-3d65c7881661)

Hope this answer helped.

**Reply by GaryAustin1** [reply]
The 1000 limit is for your protection on wasted downloads from a bad query.  You can set that higher or lower but it applies to all returns from queries.

**Reply by truexor** [reply]
thanks for the insight

**Reply by nguyenthanhtien** [reply]
No, i want ask, if i have logic like Order(check inventory, add item to cart, add promotion, payment...), supabase can support complex product with good performance or not?

**Reply by GaryAustin1** [reply]
There is no way you will get a clear answer on that.  Supabase is mainly a Postgres database under the hood with added features like storage/auth/edge functions and realtime.  Can Postgres handle your needs with its capabilities?  It has alot of ability to run business logic, but most users still run logic on client code and server code outside of the database.  

I have no idea what CCU is without looking it up.  Without saying what you need the database to do you limit those who might respond.   

But in the end it comes down to can Postgres do what you want?  It is a major database platform used by many large companies.  It can scale to handle tens of millions of records or more both vertically and horizontally.   There is lots of info on Postgres on the web.

You will mainly get a shot at other users responding here so you need to be specific in your questions.

**Comment by Rubenburdin** [comment]
Hey @nguyenthanhtien, to answer your actual question about complex order logic: yes, Supabase can handle it, but how you structure it matters a lot for 100k CCU.For transactional flows like orders (inventory check > cart > promotion > payment), you have a few options:Database functions are your friend for operations that need atomicity. An order placement that checks inventory and decrements stock should happen in a single transaction to avoid race conditions:

CREATE OR REPLACE FUNCTION place_order(p_user_id uuid, p_items jsonb)
RETURNS jsonb AS $$
DECLARE
  v_order_id uuid;
BEGIN
  -- Lock and check inventory
  -- Create order record
  -- Decrement stock
  -- Apply promotions
  -- All in one transaction
  RETURN jsonb_build_object('order_id', v_order_id);
END;
$$ LANGUAGE plpgsql;

Edge Functions work well for logic that needs external calls (payment gateways, shipping APIs) or complex validation that's awkward in SQL.
The 1000 row limit that @kush-ks mentioned is just a default safety cap on PostgREST responses - it's configurable in your project settings and has nothing to do with CCU or performance limits.

For 100k CCU specifically, the bottleneck is usually connection pooling, not business logic. Supabase uses PgBouncer, but at that scale you'll want to be strategic about when you're hitting the database directly vs using realtime subscriptions vs caching. The self-hosted AKS deployment you mentioned gives you more control over scaling the connection pool.
One pattern I'd avoid: putting all your business logic in client-side JavaScript. Fine for prototypes, but at scale you want critical operations (anything touching money or inventory) server-side where you control the execution.
